File: 2022/02/23/validAnagram.ts
/**
 * https://leetcode.com/problems/valid-anagram/
 * @author Andrew Jarombek
 * @since 2/23/2022
 */

function isAnagram(s: string, t: string): boolean {
    let chars = new Map<string, number>();

    for (let c of s) {
        chars.set(c, (chars.get(c) ?? 0) + 1);
    }

    for (let c of t) {
        let count = chars.get(c);

        if (!count) {
            return false;
        }

        if (count == 1) {
            chars.delete(c);
        } else {
            chars.set(c, count - 1);
        }
    }

    return chars.size == 0;
}

File: 2022/07/10/java/ValidAnagram.java
/**
 * https://leetcode.com/problems/valid-anagram/
 * @author Andrew Jarombek
 * @since 7/10/2022
 */

import java.util.HashMap;
import java.util.Map;

public class ValidAnagram {
    public static boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) {
            return false;
        }

        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            map.put(c, map.getOrDefault(c, 0) + 1);
        }

        for (int i = 0; i < t.length(); i++) {
            char c = t.charAt(i);
            int count = map.getOrDefault(c, 0);

            if (count == 0) {
                return false;
            }

            map.put(c, count - 1);
        }

        for (int value : map.values()) {
            if (value > 0) {
                return false;
            }
        }

        return true;
    }

    public static void main(String[] args) {
        assert !isAnagram("cat", "rat");
        assert isAnagram("dog", "god");
    }
}
